File: src/lib/topology.ts
/**
 * Plan 13 Phase 1.4 — Frontend wrapper for the topology Tauri commands.
 *
 * This module is intentionally thin: it owns the TS types for the
 * topology data model (kept in sync with `src-tauri/src/topology/mod.rs`)
 * and exposes a typed `topologyApi` that just forwards to `invoke()`.
 * Business logic (caching, selectors, error handling) lives in
 * `src/state/topologyStore.ts`.
 */
import { invoke } from "@tauri-apps/api/core";

export type Protocol = "cdp" | "lldp" | "bgp" | "ospf" | "isis";

export interface TopologyNode {
  graph_id: string;
  device_ref: string;
  device_kind: "ssh" | "netconf" | "discovered";
  label: string;
  vendor?: string;
  platform?: string;
  mgmt_ip?: string;
}

export interface TopologyEdge {
  graph_id: string;
  a_device_ref: string;
  a_port: string;
  b_device_ref: string;
  b_port: string;
  protocol: Protocol;
  captured_at: number;
}

export interface TopologyGraph {
  id: string;
  name: string;
  description?: string;
  created_at: number;
  updated_at: number;
}

export interface GraphSnapshot {
  graph: TopologyGraph;
  nodes: TopologyNode[];
  edges: TopologyEdge[];
}

export interface IngestSummary {
  nodes_added: number;
  edges_added: number;
  cache_updated: boolean;
}

export interface NeighborCacheRow {
  device_ref: string;
  device_kind: string;
  source_cmd: string;
  captured_at: number;
  parsed_json: string;
}

export const topologyApi = {
  ingestFromBlock: (
    blockId: string,
    vendor: string,
    platform: string,
    deviceRef: string,
    deviceKind: string,
  ) =>
    invoke<IngestSummary>("topology_ingest_from_block", {
      blockId,
      vendor,
      platform,
      deviceRef,
      deviceKind,
    }),
  ingestFromText: (
    command: string,
    output: string,
    vendor: string,
    platform: string,
    deviceRef: string,
    deviceKind: string,
  ) =>
    invoke<IngestSummary>("topology_ingest_from_text", {
      command,
      output,
      vendor,
      platform,
      deviceRef,
      deviceKind,
    }),
  discoverDevice: (connectionId: string, vendor: string, platform: string, password?: string) =>
    invoke<IngestSummary>("topology_discover_device", {
      connectionId,
      vendor,
      platform,
      password,
    }),
  listGraphs: () => invoke<TopologyGraph[]>("topology_list_graphs"),
  getGraph: (graphId: string) =>
    invoke<GraphSnapshot>("topology_get_graph", { graphId }),
  createGraph: (name: string, description?: string) =>
    invoke<TopologyGraph>("topology_create_graph", { name, description }),
  deleteGraph: (graphId: string) =>
    invoke<void>("topology_delete_graph", { graphId }),
  clearGraph: (graphId: string) =>
    invoke<void>("topology_clear_graph", { graphId }),
  neighborCacheList: () =>
    invoke<NeighborCacheRow[]>("topology_neighbor_cache_list"),
};

/**
 * Stable id of the seeded "Global" topology graph, mirrored from the
 * Rust constant in `src-tauri/src/topology/mod.rs`. Migration `V0040`
 * inserts the row, so the frontend always has at least one graph.
 */
export const GLOBAL_GRAPH_ID = "00000000-0000-0000-0000-0000000000g1";

/**
 * Regex for the show commands that trigger topology ingest. Phase 1
 * shipped CDP and LLDP; Phase 5 Task 5.2 extended coverage to routing
 * adjacencies (BGP / OSPF / IS-IS).
 */
export const TOPOLOGY_TRIGGER_CMD_RE =
  /^\s*show\s+(?:(?:cdp|lldp)\s+neigh(?:bor)?|ip\s+bgp\s+summ(?:ary)?|ip\s+ospf\s+neigh(?:bor)?|isis\s+neigh(?:bors)?)/i;

/**
 * Map a topology-trigger command to the protocol the sidecar should
 * normalize against. Returns `null` for non-matching commands so the
 * block-completion hook can short-circuit.
 */
export function detectTopologyProtocol(command: string): Protocol | null {
  const cmd = command.trim().toLowerCase();
  if (/^show\s+cdp\s+neigh/.test(cmd)) return "cdp";
  if (/^show\s+lldp\s+neigh/.test(cmd)) return "lldp";
  if (/^show\s+ip\s+bgp\s+summ/.test(cmd)) return "bgp";
  if (/^show\s+ip\s+ospf\s+neigh/.test(cmd)) return "ospf";
  if (/^show\s+isis\s+neigh/.test(cmd)) return "isis";
  return null;
}

/** localStorage key for the user-facing toggle to disable topology auto-ingest. */
export const ENABLE_TOPOLOGY_INGEST_KEY = "ccie:enableTopologyIngest";

/**
 * Default ON; flip to "false" in localStorage to disable. We deliberately
 * use localStorage (not a settingsStore) because there is no shared
 * settings store in this codebase yet and a single flag does not warrant
 * creating one. Any value that is not the literal string "false" leaves
 * ingest enabled.
 */
export function isTopologyIngestEnabled(): boolean {
  try {
    return window.localStorage.getItem(ENABLE_TOPOLOGY_INGEST_KEY) !== "false";
  } catch {
    return true;
  }
}

// ---------------------------------------------------------------------------
// Phase 3.1 — openNeighbor resolver
// ---------------------------------------------------------------------------

/**
 * Result of `device_lookup_by_ref` (Tauri command landing in Phase 3.2).
 * Mirrors the Rust enum that distinguishes saved SSH connections from
 * saved NETCONF devices.
 */
export interface SavedDeviceLookup {
  kind: "ssh" | "netconf";
  device_ref: string;
}

/**
 * Callback bundle for `openNeighbor`. The resolver is intentionally
 * decoupled from the UI layer — there is no shared `uiStore` in this
 * codebase, so the caller (e.g., InlineTopologyPanel) wires modal
 * visibility and tab spawning via these handlers.
 */
export interface OpenNeighborHandlers {
  /** Called when neighbor resolves to a saved SSH connection. */
  onOpenSshTab: (ref: string) => Promise<void> | void;
  /** Called when neighbor resolves to a saved NETCONF device. */
  onOpenNetconfTab: (ref: string) => Promise<void> | void;
  /** Called when neighbor is unknown — caller surfaces SaveNeighborModal. */
  onUnknownNeighbor: (node: TopologyNode) => void;
}

/**
 * Resolve a clicked topology neighbor node to a saved device and open
 * the appropriate tab. Falls back to `onUnknownNeighbor` (which the
 * caller wires to a SaveNeighborModal) when no match is found or when
 * the lookup command is unavailable.
 *
 * Lookup precedence: `node.mgmt_ip` if present, else `node.device_ref`.
 *
 * Connects directly to the resolved host. (Session chaining / jump-host
 * routing was removed in favor of Fan-Out; there is no chain resolve step.)
 *
 * Plan 13 spec (lines 549–582) explicitly allows this direct-connect path.
 */
export async function openNeighbor(
  node: TopologyNode,
  handlers: OpenNeighborHandlers,
): Promise<void> {
  const ref = node.mgmt_ip ?? node.device_ref;
  let saved: SavedDeviceLookup | null = null;
  try {
    saved = await invoke<SavedDeviceLookup | null>("device_lookup_by_ref", {
      ref,
    });
  } catch (err) {
    // device_lookup_by_ref may not be wired yet (Task 3.2 lands in
    // parallel). Treat any failure as "unknown neighbor" so the caller
    // can offer the SaveNeighborModal flow.
    console.warn("[topology] device_lookup_by_ref unavailable:", err);
  }

  if (!saved) {
    handlers.onUnknownNeighbor(node);
    return;
  }

  if (saved.kind === "ssh") {
    await handlers.onOpenSshTab(saved.device_ref);
    return;
  }

  if (saved.kind === "netconf") {
    await handlers.onOpenNetconfTab(saved.device_ref);
    return;
  }
}

File: src-tauri/src/commands/topology.rs
//! Plan 13 Phase 1.3 — Tauri commands exposing the topology module.
//!
//! Frontend integration:
//!   - When a `show cdp/lldp neighbor*` block completes, the frontend
//!     calls `topology_ingest_from_block(blockId, vendor, platform,
//!     deviceRef, deviceKind)`. Vendor/platform live on the Tab in
//!     in-memory frontend state; the same pattern is used by
//!     `structured_auto_parse`.
//!   - The other commands are CRUD wrappers used by the Topology tab and
//!     the inline topology panel.

use regex::Regex;
use rusqlite::params;
use std::sync::OnceLock;
use tauri::State;

use super::AppState;
use crate::topology::{
    cache, graph, ingest, lookup, GraphSnapshot, IngestSummary, NeighborCacheRow,
    SavedDeviceLookup, TopologyGraph,
};

/// Captures the protocol token from a topology-trigger show command.
///
/// Phase 1 covered CDP/LLDP; Phase 5 Task 5.2 extended to routing
/// adjacencies (BGP / OSPF / IS-IS).
fn protocol_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| {
        Regex::new(
            r"(?i)^\s*show\s+(?:(?P<cdp>cdp)\s+neigh(?:bor)?|(?P<lldp>lldp)\s+neigh(?:bor)?|ip\s+(?P<bgp>bgp)\s+summ(?:ary)?|ip\s+(?P<ospf>ospf)\s+neigh(?:bor)?|(?P<isis>isis)\s+neigh(?:bors)?)",
        )
        .expect("topology protocol regex must compile")
    })
}

fn detect_protocol(cmd: &str) -> Option<String> {
    let caps = protocol_re().captures(cmd)?;
    for name in ["cdp", "lldp", "bgp", "ospf", "isis"] {
        if caps.name(name).is_some() {
            return Some(name.to_string());
        }
    }
    None
}

/// Read `(cmd, output)` for a block. `output` is BLOB in the schema; we
/// decode lossily because show-output is virtually always UTF-8 ANSI.
fn read_block(conn: &rusqlite::Connection, block_id: &str) -> anyhow::Result<(String, String)> {
    let row: (String, Vec<u8>) = conn.query_row(
        "SELECT cmd, output FROM command_blocks WHERE id = ?1",
        params![block_id],
        |r| Ok::<(String, Vec<u8>), rusqlite::Error>((r.get(0)?, r.get(1)?)),
    )?;
    let output = String::from_utf8_lossy(&row.1).into_owned();
    Ok((row.0, output))
}

/// Parse + ingest neighbors from the given completed block.
#[tauri::command]
pub async fn topology_ingest_from_block(
    state: State<'_, AppState>,
    block_id: String,
    vendor: String,
    platform: String,
    device_ref: String,
    device_kind: String,
) -> Result<IngestSummary, String> {
    // 1. Read the block's cmd + output (sync; release the lock before sidecar I/O).
    let (cmd, raw_output) = {
        let conn = state.db.lock();
        read_block(&conn, &block_id).map_err(|e| e.to_string())?
    };

    // 2. Detect cdp vs lldp from the command string.
    let protocol = detect_protocol(&cmd).ok_or_else(|| {
        format!(
            "topology_ingest_from_block: command does not match a supported topology trigger (cdp/lldp/bgp/ospf/isis): {cmd}"
        )
    })?;

    // 3. Call the sidecar to normalize the output.
    let records = state
        .parser_bridge
        .neighbors(&protocol, &vendor, &platform, &cmd, &raw_output)
        .await
        .map_err(|e| e.to_string())?;

    // 4. Persist into neighbor_cache + topology_{nodes,edges}.
    let summary = {
        let conn = state.db.lock();
        ingest::ingest_neighbors(
            &conn,
            &device_ref,
            &device_kind,
            &cmd,
            &raw_output,
            &records,
        )
        .map_err(|e| e.to_string())?
    };

    Ok(summary)
}

/// Manual topology ingestion from raw text (for Terminal mode SSH sessions).
///
/// This command allows users to manually trigger topology ingestion without
/// requiring a block ID. Useful for SSH sessions in Terminal mode where
/// automatic block-based ingestion doesn't work.
#[tauri::command]
pub async fn topology_ingest_from_text(
    state: State<'_, AppState>,
    command: String,
    output: String,
    vendor: String,
    platform: String,
    device_ref: String,
    device_kind: String,
) -> Result<IngestSummary, String> {
    // 1. Detect protocol from the command string.
    let protocol = detect_protocol(&command).ok_or_else(|| {
        format!(
            "topology_ingest_from_text: command does not match a supported topology trigger (cdp/lldp/bgp/ospf/isis): {command}"
        )
    })?;

    // 2. Call the sidecar to normalize the output.
    let records = state
        .parser_bridge
        .neighbors(&protocol, &vendor, &platform, &command, &output)
        .await
        .map_err(|e| e.to_string())?;

    // 3. Persist into neighbor_cache + topology_{nodes,edges}.
    let summary = {
        let conn = state.db.lock();
        ingest::ingest_neighbors(
            &conn,
            &device_ref,
            &device_kind,
            &command,
            &output,
            &records,
        )
        .map_err(|e| e.to_string())?
    };

    Ok(summary)
}

#[tauri::command]
pub fn topology_list_graphs(state: State<'_, AppState>) -> Result<Vec<TopologyGraph>, String> {
    let conn = state.db.lock();
    graph::list_graphs(&conn).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn topology_get_graph(
    state: State<'_, AppState>,
    graph_id: String,
) -> Result<GraphSnapshot, String> {
    let conn = state.db.lock();
    let g = graph::get_graph(&conn, &graph_id)
        .map_err(|e| e.to_string())?
        .ok_or_else(|| format!("topology graph not found: {graph_id}"))?;
    let nodes = graph::list_nodes(&conn, &graph_id).map_err(|e| e.to_string())?;
    let edges = graph::list_edges(&conn, &graph_id).map_err(|e| e.to_string())?;
    Ok(GraphSnapshot {
        graph: g,
        nodes,
        edges,
    })
}

#[tauri::command]
pub fn topology_create_graph(
    state: State<'_, AppState>,
    name: String,
    description: Option<String>,
) -> Result<TopologyGraph, String> {
    let id = uuid::Uuid::new_v4().to_string();
    let conn = state.db.lock();
    graph::create_graph(&conn, &id, &name, description.as_deref()).map_err(|e| e.to_string())?;
    graph::get_graph(&conn, &id)
        .map_err(|e| e.to_string())?
        .ok_or_else(|| "topology_create_graph: row vanished immediately after insert".to_string())
}

#[tauri::command]
pub fn topology_delete_graph(state: State<'_, AppState>, graph_id: String) -> Result<(), String> {
    let conn = state.db.lock();
    graph::delete_graph(&conn, &graph_id).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn topology_clear_graph(state: State<'_, AppState>, graph_id: String) -> Result<(), String> {
    let conn = state.db.lock();
    graph::clear_graph(&conn, &graph_id).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn topology_neighbor_cache_list(
    state: State<'_, AppState>,
) -> Result<Vec<NeighborCacheRow>, String> {
    let conn = state.db.lock();
    cache::list_recent(&conn, 256).map_err(|e| e.to_string())
}

/// Plan 13 Phase 3.2 — resolve a clicked topology neighbor to a saved
/// SSH connection or NETCONF device.
///
/// The frontend (`src/lib/topology.ts::openNeighbor`) invokes this with
/// `{ ref }` (the camelCase JS identifier `ref`). Because `ref` is a
/// reserved Rust keyword, we use the raw-identifier form `r#ref` here;
/// Tauri's serde layer maps the JSON field `ref` to it transparently.
///
/// SSH is the primary interactive path; when a host is saved in BOTH
/// tables we prefer SSH so click-to-SSH stays predictable. See
/// `topology::lookup::lookup_device_by_ref` for the underlying SQL.
#[tauri::command]
pub fn device_lookup_by_ref(
    state: State<'_, AppState>,
    r#ref: String,
) -> Result<Option<SavedDeviceLookup>, String> {
    let conn = state.db.lock();
    lookup::lookup_device_by_ref(&conn, &r#ref).map_err(|e| e.to_string())
}

/// Automated topology discovery: SSH to a device and run neighbor discovery commands.
///
/// This command:
/// 1. Looks up the saved SSH connection by ID
/// 2. Connects via SSH using russh
/// 3. Runs 'show cdp neighbors detail' and 'show lldp neighbors detail'
/// 4. Calls topology_ingest_from_text for each command
/// 5. Returns summary of neighbors discovered
#[tauri::command]
pub async fn topology_discover_device(
    state: State<'_, AppState>,
    connection_id: String,
    vendor: String,
    platform: String,
    password: Option<String>,
) -> Result<IngestSummary, String> {
    // 1. Resolve the saved SSH connection into a target (decrypts/overrides
    //    password). Shared with the other SSH-direct features.
    let (target, device_name) =
        crate::ssh_exec::resolve_target(&state.db, &connection_id, password)?;

    // 2. Run discovery commands via the shared sshpass-backed executor.
    //    russh has limited algorithm support; the system ssh client works
    //    with more devices.
    let commands = [
        ("cdp", "show cdp neighbors detail"),
        ("lldp", "show lldp neighbors detail"),
    ];

    let mut total_nodes = 0;
    let mut total_edges = 0;

    for (protocol, cmd) in commands {
        let raw_output =
            match crate::ssh_exec::run_command(&target, cmd, crate::ssh_exec::DEFAULT_CMD_TIMEOUT)
                .await
            {
                Ok(out) => out,
                Err(e) => {
                    eprintln!("{} command failed: {}", protocol.to_uppercase(), e);
                    continue; // Try next protocol even if this one fails
                }
            };

        // Parse and ingest this protocol
        match state
            .parser_bridge
            .neighbors(protocol, &vendor, &platform, cmd, &raw_output)
            .await
        {
            Ok(records) => {
                let conn = state.db.lock();
                match ingest::ingest_neighbors(
                    &conn,
                    &device_name,
                    "ssh",
                    cmd,
                    &raw_output,
                    &records,
                ) {
                    Ok(summary) => {
                        total_nodes += summary.nodes_added;
                        total_edges += summary.edges_added;
                    }
                    Err(e) => eprintln!("{} ingest failed: {}", protocol.to_uppercase(), e),
                }
            }
            Err(e) => eprintln!("{} parse failed: {}", protocol.to_uppercase(), e),
        }
    }

    Ok(IngestSummary {
        nodes_added: total_nodes,
        edges_added: total_edges,
        cache_updated: true,
    })
}
